cub/test/catch2_test_device_find_if.cu

```diff
@@ -0,0 +1,237 @@
+using full_type_list = c2h::type_list<type_pair<std::int32_t>, type_pair<std::int64_t, std::int32_t>>;
+#elif TEST_TYPES == 2
+using full_type_list =
+  c2h::type_list<type_pair<uchar3, std::int32_t>,
+                 type_pair<
+#  if _CCCL_CTK_AT_LEAST(13, 0)
+                   ulonglong4_16a,
+#  else // _CCCL_CTK_AT_LEAST(13, 0)
+                   ulonglong4,
+#  endif // _CCCL_CTK_AT_LEAST(13, 0)
+                   std::int32_t>>;
+#elif TEST_TYPES == 3
+// clang-format off
+using full_type_list = c2h::type_list<
+type_pair<custom_t, std::int32_t>
+#if TEST_HALF_T()
+, type_pair<half_t, std::int32_t> // testing half
+#endif // TEST_HALF_T()
+#if TEST_BF_T()
+, type_pair<bfloat16_t, std::int32_t> // testing bf16
+#endif // TEST_BF_T()
+>;
+// clang-format on
+#endif
+
+enum class gen_data_t : int
+{
+  /// Uniform random data generation
+  GEN_TYPE_RANDOM,
+  /// Constant value as input data
+  GEN_TYPE_CONST
+};
+
+template <typename InputIt, typename OutputIt, typename BinaryOp>
+void compute_find_if_reference(InputIt first, InputIt last, OutputIt& result, BinaryOp op)
+{
+  auto pos = thrust::find_if(first, last, op);
+  result   = static_cast<OutputIt>(pos - first);
+}
+
+template <typename T>
+struct equals
+{
+  T val;
+
+  __device__ __host__ bool operator()(T i) const
+  {
+    return i == val;
+  }
+
+  // Accept any type convertible to T (for half_t/__half, bfloat16_t/__nv_bfloat16 compatibility)
+  template <typename U, typename = typename ::cuda::std::enable_if<!::cuda::std::is_same<T, U>::value>::type>
+  __device__ __host__ bool operator()(U i) const
+  {
+    return T(i) == val;
+  }
+};
+
+C2H_TEST("Device find_if works", "[device]", full_type_list)
+{
+  using params   = params_t<TestType>;
+  using input_t  = typename params::item_t;
+  using output_t = typename params::output_t;
+  using offset_t = output_t;
+
+  constexpr offset_t min_items = 1;
+  constexpr offset_t max_items = std::numeric_limits<offset_t>::max() / 5; // make test run faster
+
+  // Generate the input sizes to test for
+  const offset_t num_items = GENERATE_COPY(
+    take(3, random(min_items, max_items)),
+    values({
+      min_items,
+      max_items,
+    }));
+
+  // Input data generation to test
+  const gen_data_t data_gen_mode = GENERATE_COPY(gen_data_t::GEN_TYPE_RANDOM, gen_data_t::GEN_TYPE_CONST);
+
+  // Generate input data
+  c2h::device_vector<input_t> in_items(num_items);
+  if (data_gen_mode == gen_data_t::GEN_TYPE_RANDOM)
+  {
+    c2h::gen(C2H_SEED(2), in_items);
+  }
+  else
+  {
+    input_t default_constant{};
+    init_default_constant(default_constant);
+    thrust::fill(c2h::device_policy, in_items.begin(), in_items.end(), default_constant);
+  }
+  auto d_in_it = thrust::raw_pointer_cast(in_items.data());
+
+  using op_t = equals<input_t>;
+  input_t val_to_find{};
+  if constexpr (::cuda::std::is_arithmetic_v<input_t>)
+  {
+    val_to_find = static_cast<input_t>(GENERATE_COPY(take(1, random(min_items, max_items))));
```

Comment:
Question: how can this work? If we generate a random value, it's not given that it actually occurs in the input sequence. It may even be very unlikely that it is.

Suggestion: I think we should have dedicated tests where the value is found and where it is not found.